skorch/net.py

```diff
@@ -1524,6 +1534,90 @@ def __setstate__(self, state):
 
         self.__dict__.update(state)
 
+    def _register_attribute(
+            self,
+            name,
+            prefixes=True,
+            cuda_dependent_attributes=True,
+    ):
+        """Add attribute name to prefixes_ and
+        cuda_dependent_attributes_.
+
+        Takes care of not mutating the lists.
+
+        Parameters
+        ----------
+        prefixes : bool (default=True)
+          Whether to add to prefixes_.
+
+        cuda_dependent_attributes : bool (default=True)
+          Whether to add to cuda_dependent_attributes_.
+
+        """
+        # copy the lists to avoid mutation
+        if prefixes:
+            self.prefixes_ = self.prefixes_[:] + [name]
+
+        if cuda_dependent_attributes:
+            self.cuda_dependent_attributes_ = (
+                self.cuda_dependent_attributes_[:] + [name + '_'])
+
+    def _unregister_attribute(
+            self,
+            name,
+            prefixes=True,
+            cuda_dependent_attributes=True,
+    ):
+        """Remove attribute name from prefixes_ and
+        cuda_dependent_attributes_.
+
+        Takes care of not mutating the lists.
+
+        Parameters
+        ----------
+        prefixes : bool (default=True)
+          Whether to remove from prefixes_.
+
+        cuda_dependent_attributes : bool (default=True)
+          Whether to remove from cuda_dependent_attributes_.
+
+        """
+        # copy the lists to avoid mutation
+        if prefixes:
+            self.prefixes_ = [p for p in self.prefixes_[:] if p != name]
+
+        if cuda_dependent_attributes:
+            self.cuda_dependent_attributes_ = [
+                a for a in self.cuda_dependent_attributes_ if a != name + '_']
+
+    def __setattr__(self, name, attr):
+        """Set an attribute on the net
+
+        When a custom net with additional torch modules or optimizers
+        is created, those attributes are added to ``prefixes_`` and
+        ``cuda_dependent_attributes_`` automatically.
+
+        """
+        # if it's a
+        # 1. known attribute or
+        # 2. special param like module__num_units or
+        # 3. not a torch module/optimizer instance or class
+        # just setattr as usual
+        is_known = name.endswith('_') or (name in self.prefixes_)
+        is_special_param = '__' in name
+        is_torch_mod_or_opt = any(c in name for c in PYTORCH_COMPONENTS)
+        if is_known or is_special_param or not is_torch_mod_or_opt:
+            super().__setattr__(name, attr)
+            return
+
+        self._register_attribute(name)
+        super().__setattr__(name, attr)
```

Comment:
Wouldn't it be more straight forward to reduce this to

1. register if necessary
2. set attribute

instead of having (check, set) and (register, set)?

Suggestion:

```suggestion
if is_torch_mod_or_opt or not (is_known and is_special_param):
    self._register_attribute(name)
    
super().__setattr__(name, attr)
```

Reply:
Streamlined the code (though a bit different than proposed)